### PR TITLE
Speedup reading of results

### PR DESCRIPTION
**Pull request by daniel-thom**
This PR refactors the storage and access of results in order to improve performance. DataFrames are replaced by DenseAxisArrays. It also optimizes the generation of realized results.

Removed a lot of duplication in these methods and underlying code. Callers should now use `read_results_with_keys` instead.
```
read_aux_variables_with_keys
read_duals_with_keys
read_expressions_with_keys
read_parameters_with_keys
read_variables_with_keys
```

### CODE REVIEW

File: src/core/dataset.jl [hunk 1]
```diff
@@ -150,6 +167,8 @@ function get_last_updated_timestamp(s::HDF5Dataset)
 end
 
 function get_value_timestamp(s::HDF5Dataset, date::Dates.DateTime)
+    # TODO: This code is broken because timestamps is not a field.
```
[the remark this answers is not in the document]
Comment (jd-lara) [reply src/core/dataset.jl:170]: I am wondering why this function isn't tested.

[the remark this answers is not in the document]
Comment (jd-lara) [reply src/core/dataset.jl:170]: Let's add an error that just says something along the lines of "not implemented for `HDF5Dataset`. Required only if HDF5Dataset is used for simulation state"

File: src/core/results_by_time.jl [hunk 2]
```diff
@@ -0,0 +1,84 @@
+struct ResultsByTime{T}
+    key::OptimizationContainerKey
+    data::SortedDict{Dates.DateTime, T}
+    resolution::Dates.Period
+    column_names::Vector{String}
+end
+
+function ResultsByTime(key, data, resolution, column_names)
```
Comment (jd-lara) [inline src/core/results_by_time.jl:8]:
```suggestion
function ResultsByTime(key, data, resolution, column_names::Vector{String})
```

File: src/core/results_by_time.jl [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    _check_column_consistency(data, column_names)
+    ResultsByTime(key, data, resolution, column_names)
+end
+
+function _check_column_consistency(
+    data::SortedDict{Dates.DateTime, DenseAxisArray{Float64, 2}},
+    cols,
```
Comment (jd-lara) [inline src/core/results_by_time.jl:15]:
```suggestion
    cols::Vector{String},
```

File: src/operation/problem_results.jl [hunk 4]
```diff
@@ -297,7 +303,7 @@ function _read_results(
     container_keys,
     timestamps,
```
Comment (jd-lara) [inline src/operation/problem_results.jl:304]: check typing here

File: src/operation/problem_results.jl [hunk 5]
```diff
@@ -297,7 +303,7 @@ function _read_results(
...
     time_ids,
-    base_power,
+    base_power::Number,
```
Comment (jd-lara) [inline src/operation/problem_results.jl:306]: This is always a Float64

File: src/operation/problem_results.jl [hunk 6]
```diff
@@ -764,27 +706,28 @@ function read_expressions(
...
 """
 Return the values for all expressions.
 """
 function read_expressions(res::IS.Results)
-    expressions = Dict(x => read_expression(res, x) for x in list_expression_names(res))
+    return Dict(x => read_expression(res, x) for x in list_expression_names(res))
 end
 
-function read_expressions_with_keys(
+function read_results_with_keys(
```
Comment (jd-lara) [inline src/operation/problem_results.jl:720]: Add note that the existence of this function might need to be reviewed

File: src/parameters/update_parameters.jl [hunk 7]
```diff
@@ -166,7 +166,7 @@ function _update_parameter_values!(
     attributes::VariableValueAttributes,
     ::Type{<:PSY.Component},
     model::DecisionModel,
-    state::DatasetContainer{DataFrameDataset},
+    state::DatasetContainer{InMemoryDataset},
```
Comment (jd-lara) [inline src/parameters/update_parameters.jl:169]: This function needs to be tested or check a case where this is called. Potentially delete or comment out.

File: src/parameters/update_parameters.jl [hunk 8]
```diff
@@ -268,7 +268,7 @@ function _update_parameter_values!(
     attributes::VariableValueAttributes{VariableKey{OnVariable, U}},
     ::Type{<:PSY.Component},
     model::EmulationModel,
```
Comment (jd-lara) [inline src/parameters/update_parameters.jl:270]: Might need test with an emulation model that uses Unit Commitment.

File: src/simulation/emulation_model_simulation_results.jl [hunk 9]
```diff
@@ -150,40 +210,33 @@ function _check_offsets(
...
-    res::SimulationProblemResults{EmulationModelSimulationResults},
-    aux_variables::Vector{<:OptimizationContainerKey};
-    start_time::Union{Nothing, Dates.DateTime} = nothing,
-    len::Union{Nothing, Int} = nothing,
-)
-    return _read_aux_variables(res, aux_variables, nothing, start_time, len)
-end
-
-function _read_aux_variables(
+function _read_results(
```
Comment (jd-lara) [inline src/simulation/emulation_model_simulation_results.jl:213]: review that we should type inputs to the function here

File: src/simulation/decision_model_simulation_results.jl [hunk 10]
```diff
@@ -139,27 +218,28 @@ function _process_timestamps(
     return requested_range
 end
 
-function _read_variables(
+function _read_results(
+    ::Type{T},
     res::SimulationProblemResults{DecisionModelSimulationResults},
-    variable_keys,
+    result_keys,
```
Comment (jd-lara) [inline src/simulation/decision_model_simulation_results.jl:224]: This can be typed

File: src/simulation/hdf_simulation_store.jl [hunk 11]
```diff
@@ -345,7 +345,7 @@ function read_result(
     if (ndims(data) < 2 || size(data)[1] == 1) && size(data)[2] != size(columns)[1]
         data = reshape(data, length(data), 1)
     end
-    return DataFrames.DataFrame(data, columns)
+    return DataFrames.DataFrame(data, collect(columns); copycols = false)
```
Comment (jd-lara) [inline src/simulation/hdf_simulation_store.jl:348]: Check if the collect is needed.

File: src/simulation/hdf_simulation_store.jl [hunk 12]
```diff
@@ -492,10 +519,9 @@ function write_result!(
     key::OptimizationContainerKey,
     index::DecisionModelIndexType,
     ::Dates.DateTime,
-    data,
-)
+    data::DenseAxisArray{Float64, N, K},
```
Comment (jd-lara) [inline src/simulation/hdf_simulation_store.jl:522]: @daniel-thom might get rid of the K

File: src/utils/dataframes_utils.jl [hunk 13]
```diff
@@ -7,20 +7,20 @@ Creates a DataFrame from a JuMP DenseAxisArray or SparseAxisArray.
...
     return DataFrames.DataFrame(to_matrix(array), get_column_names(key, array))
 end
 
-function axis_array_to_dataframe(array::SparseAxisArray, key::OptimizationContainerKey)
+function to_dataframe(array::SparseAxisArray, key::OptimizationContainerKey)
     return DataFrames.DataFrame(to_matrix(array), get_column_names(key, array))
 end
 
-function axis_array_to_dataframe(array::Matrix{Float64}, key::OptimizationContainerKey)
+function to_dataframe(array::Matrix{Float64}, key::OptimizationContainerKey)
```
Comment (jd-lara) [inline src/utils/dataframes_utils.jl:18]: we should remove this. This function shouldn't exist and doesn't work

**Review by jd-lara — CHANGES_REQUESTED** [review]
@daniel-thom and I conducted a review and some changes are recommended.

File: src/operation/problem_results.jl (hunk 6 above)
Comment (daniel-thom) [reply src/operation/problem_results.jl:720]: I’m pretty sure the reason is that Clayton wanted to be able to treat problems and simulation problems the same in PowerGraphics.

File: src/PowerSimulations.jl [hunk 14]
```diff
@@ -575,4 +576,12 @@ include("utils/recorder_events.jl")
 include("utils/datetime_utils.jl")
 include("utils/generate_valid_formulations.jl")
 
+# TODO: These exist for backward compatibility and need to be deprecated and removed.
```
[the remark this answers is not in the document]
Comment (jd-lara) [reply src/PowerSimulations.jl:579]: agreed

File: src/operation/problem_results.jl (hunk 6 above)
Comment (jd-lara) [reply src/operation/problem_results.jl:720]: ok, lets leave it for now

**Review by jd-lara — APPROVED** [review]

**Pull request merged**